### PR TITLE
Replace rendezvous with a faster atomic barrier

### PR DESCRIPTION
**Pull request by dsunder**
The rendezvous algorithm relied on volatile int8_t being written atomically to avoid explicit atomic operations.  However, by avoiding atomic operations the compiler did not prioritize those writes.  This interacted poorly with our spinwait algorithm which uses an exponential backoff strategy.  Once a thread yielded or slept there were instances that a thread never completed its write before its quantum expired resulting in a live-lock.  This effect was amplified when threads were unbounded, oversubscribed, or the process touched multiple numa regions.  Using an atomic operation guarantees that a thread will complete its write/load before yielding the quantum.

Since the std::thread backend heavily relied on the rendezvous algorithm, it would fail more frequently.  Much time was spent avoiding thread local storage, cleaning up scratch memory, and trying to harden the rendezvous algorithm.  However, I now believe that those efforts were attempts to work around a fundamental flaw in the rendezvous algorithm, namely that it did not provide a forward progress guarantee.

As a result of this change the memory fences are reverted to being the fastest assembly and/or compiler builtins available.

My current result seem to indicate that the new barrier is faster than the rendezvous algorithm, but that is only using the eyeball metric.  The new barrier is significantly faster (and safer) when threads are oversubscribed.

I would recommend viewing the entire file when reviewing Impl/Kokkos_HostBarrier.* since it is entirely new content.

### CODE REVIEW

**Comment by mhoemmen** [comment]
@bartlettroscoe @nmhamster @micahahoward @prwolfe @rrdrake

This change may mitigate some of the effects of oversubscription during Trilinos(, Sierra, etc.) testing with OpenMP.  That could decrease the risks of testing OpenMP by default, though it would still behoove testers as well as users to avoid oversubscription.

**Comment by bartlettroscoe** [comment]
> This change may mitigate some of the effects of oversubscription during Trilinos(, Sierra, etc.) testing with OpenMP. 

@mhoemmen, are you referring to #2398?  If so, changing the mpiexec options seems to have fixed that.  I just closed that issue this morning.

**Comment by mhoemmen** [comment]
@bartlettroscoe I'm thinking more in general, but thanks for fixing that :-) .

File: core/src/impl/Kokkos_HostBarrier.hpp [hunk 1]
```diff
@@ -75,7 +75,9 @@ namespace Kokkos { namespace Impl {
 class HostBarrier
 {
 public:
-  static constexpr int required_buffer_size = 128;
+  using buffer_type = int;
```
Comment (dalg24) [inline core/src/impl/Kokkos_HostBarrier.hpp]: `buffer_type` does not seem to be used anywhere.

**Comment by dsunder** [comment]
kokkos-dev
#######################################################
PASSED TESTS
#######################################################
clang-3.6.1-Pthread-hwloc-release build_time=182 run_time=250
clang-3.6.1-Pthread-release build_time=176 run_time=365
clang-3.6.1-Pthread_Serial-hwloc-release build_time=207 run_time=201
clang-3.6.1-Pthread_Serial-release build_time=150 run_time=356
clang-3.6.1-Serial-hwloc-release build_time=150 run_time=176
clang-3.6.1-Serial-release build_time=148 run_time=167
clang-3.7.1-Pthread-hwloc-release build_time=218 run_time=124
clang-3.7.1-Pthread-release build_time=218 run_time=286
clang-3.7.1-Pthread_Serial-hwloc-release build_time=208 run_time=204
clang-3.7.1-Pthread_Serial-release build_time=196 run_time=369
clang-3.7.1-Serial-hwloc-release build_time=200 run_time=177
clang-3.7.1-Serial-release build_time=154 run_time=222
clang-3.8.1-Pthread-hwloc-release build_time=248 run_time=92
clang-3.8.1-Pthread-release build_time=384 run_time=228
clang-3.8.1-Pthread_Serial-hwloc-release build_time=289 run_time=196
clang-3.8.1-Pthread_Serial-release build_time=294 run_time=346
clang-3.8.1-Serial-hwloc-release build_time=254 run_time=184
clang-3.8.1-Serial-release build_time=214 run_time=204
clang-3.9.0-Pthread-hwloc-release build_time=234 run_time=86
clang-3.9.0-Pthread-release build_time=332 run_time=242
clang-3.9.0-Pthread_Serial-hwloc-release build_time=253 run_time=304
clang-3.9.0-Pthread_Serial-release build_time=262 run_time=487
clang-3.9.0-Serial-hwloc-release build_time=245 run_time=228
clang-3.9.0-Serial-release build_time=198 run_time=217
cuda-7.0.28-Cuda_OpenMP-release build_time=734 run_time=557
cuda-7.0.28-Cuda_Pthread-release build_time=755 run_time=588
cuda-7.0.28-Cuda_Serial-release build_time=637 run_time=456
cuda-7.5.18-Cuda_OpenMP-release build_time=632 run_time=465
cuda-7.5.18-Cuda_Pthread-release build_time=613 run_time=463
cuda-7.5.18-Cuda_Serial-release build_time=560 run_time=517
cuda-8.0.44-Cuda_OpenMP-release build_time=549 run_time=543
cuda-8.0.44-Cuda_Pthread-release build_time=533 run_time=528
cuda-8.0.44-Cuda_Serial-release build_time=518 run_time=508
gcc-4.8.4-OpenMP-hwloc-release build_time=211 run_time=424
gcc-4.8.4-OpenMP-release build_time=208 run_time=422
gcc-4.8.4-OpenMP_Serial-hwloc-release build_time=227 run_time=211
gcc-4.8.4-OpenMP_Serial-release build_time=205 run_time=388
gcc-4.8.4-Pthread-hwloc-release build_time=160 run_time=113
gcc-4.8.4-Pthread-release build_time=190 run_time=548
gcc-4.8.4-Pthread_Serial-hwloc-release build_time=174 run_time=206
gcc-4.8.4-Pthread_Serial-release build_time=192 run_time=525
gcc-4.8.4-Serial-hwloc-release build_time=185 run_time=518
gcc-4.8.4-Serial-release build_time=144 run_time=395
gcc-4.9.3-OpenMP-hwloc-release build_time=187 run_time=140
gcc-4.9.3-OpenMP-release build_time=292 run_time=200
gcc-4.9.3-OpenMP_Serial-hwloc-release build_time=214 run_time=176
gcc-4.9.3-OpenMP_Serial-release build_time=239 run_time=187
gcc-4.9.3-Pthread-hwloc-release build_time=168 run_time=197
gcc-4.9.3-Pthread-release build_time=184 run_time=468
gcc-4.9.3-Pthread_Serial-hwloc-release build_time=194 run_time=201
gcc-4.9.3-Pthread_Serial-release build_time=204 run_time=496
gcc-4.9.3-Serial-hwloc-release build_time=166 run_time=353
gcc-4.9.3-Serial-release build_time=144 run_time=569
gcc-5.3.0-OpenMP-hwloc-release build_time=288 run_time=343
gcc-5.3.0-OpenMP-release build_time=323 run_time=165
gcc-5.3.0-OpenMP_Serial-hwloc-release build_time=333 run_time=175
gcc-5.3.0-OpenMP_Serial-release build_time=325 run_time=229
gcc-5.3.0-Pthread-hwloc-release build_time=263 run_time=144
gcc-5.3.0-Pthread-release build_time=245 run_time=336
gcc-5.3.0-Pthread_Serial-release build_time=297 run_time=227
gcc-5.3.0-Serial-hwloc-release build_time=241 run_time=231
gcc-5.3.0-Serial-release build_time=235 run_time=217
gcc-6.1.0-OpenMP-hwloc-release build_time=313 run_time=124
gcc-6.1.0-OpenMP-release build_time=316 run_time=117
gcc-6.1.0-OpenMP_Serial-hwloc-release build_time=309 run_time=178
gcc-6.1.0-OpenMP_Serial-release build_time=361 run_time=154
gcc-6.1.0-Pthread-hwloc-release build_time=238 run_time=83
gcc-6.1.0-Pthread-release build_time=275 run_time=183
gcc-6.1.0-Pthread_Serial-hwloc-release build_time=296 run_time=144
gcc-6.1.0-Pthread_Serial-release build_time=228 run_time=243
gcc-6.1.0-Serial-hwloc-release build_time=209 run_time=245
gcc-6.1.0-Serial-release build_time=209 run_time=212
intel-15.0.2-OpenMP-hwloc-release build_time=639 run_time=200
intel-15.0.2-OpenMP-release build_time=660 run_time=185
intel-15.0.2-OpenMP_Serial-hwloc-release build_time=620 run_time=233
intel-15.0.2-OpenMP_Serial-release build_time=651 run_time=211
intel-15.0.2-Pthread-hwloc-release build_time=508 run_time=173
intel-15.0.2-Pthread-release build_time=471 run_time=286
intel-15.0.2-Pthread_Serial-hwloc-release build_time=553 run_time=221
intel-15.0.2-Pthread_Serial-release build_time=473 run_time=381
intel-15.0.2-Serial-hwloc-release build_time=374 run_time=306
intel-15.0.2-Serial-release build_time=419 run_time=266
intel-16.0.1-OpenMP-hwloc-release build_time=656 run_time=230
intel-16.0.1-OpenMP-release build_time=715 run_time=193
intel-16.0.1-OpenMP_Serial-hwloc-release build_time=642 run_time=301
intel-16.0.1-OpenMP_Serial-release build_time=652 run_time=324
intel-16.0.1-Pthread-hwloc-release build_time=574 run_time=182
intel-16.0.1-Pthread-release build_time=602 run_time=213
intel-16.0.1-Pthread_Serial-hwloc-release build_time=450 run_time=225
intel-16.0.1-Pthread_Serial-release build_time=588 run_time=249
intel-16.0.1-Serial-hwloc-release build_time=476 run_time=194
intel-16.0.1-Serial-release build_time=430 run_time=262
intel-16.0.3-OpenMP-hwloc-release build_time=693 run_time=215
intel-16.0.3-OpenMP-release build_time=594 run_time=266
intel-16.0.3-OpenMP_Serial-hwloc-release build_time=778 run_time=227
intel-16.0.3-OpenMP_Serial-release build_time=680 run_time=306
intel-16.0.3-Pthread-hwloc-release build_time=450 run_time=232
intel-16.0.3-Pthread-release build_time=464 run_time=284
intel-16.0.3-Pthread_Serial-hwloc-release build_time=475 run_time=327
intel-16.0.3-Pthread_Serial-release build_time=586 run_time=278
intel-16.0.3-Serial-hwloc-release build_time=444 run_time=238
intel-16.0.3-Serial-release build_time=489 run_time=188

Also passed all tests on white and bowman

**Pull request merged**